docs/spec/concepts.rst

```diff
@@ -3,97 +3,181 @@
+if the set of values represented by ``A`` is a superset of the values
+represented by ``B``. The supertype relation is also transitive and reflexive.
+
+We also define an **equivalence** relation on static types: the types ``A`` and
+``B`` are equivalent (or "the same type") if and only if ``A`` is a subtype
+of ``B`` and ``B`` is a subtype of ``A``. This means that the set of values
+represented by ``A`` is both a superset and a subset of the values represented
+by ``B``, meaning ``A`` and ``B`` must represent the same set of values.
+
+We may describe a type ``B`` as "narrower" than a type ``A`` (or as a "strict
+subtype" of ``A``) if ``B`` is a subtype of ``A`` and ``B`` is not equivalent
+to ``A``.
+
+The consistency relation
+------------------------
+
+Since :ref:`Any` represents an unknown static type, it does not represent any
+known single set of values, and thus it is not in the domain of the subtype,
+supertype, or equivalence relations on static types described above.
+
+We define a **materialization** relation on gradual types as follows: if
+replacing zero or more occurrences of ``Any`` in gradual type ``A`` with some
+gradual type (which can be different for each occurrence of ``Any``) results in
+the gradual type ``B``, then ``B`` is a materialization of ``A``. For instance,
+``tuple[int, str]`` (a static type) and ``tuple[Any, str]`` (a gradual type)
+are both materializations of ``tuple[Any, Any]``.
+
+If ``B`` is a materialization of ``A``, we can say that ``B`` is a "more
+static" type than ``A``, and ``A`` is a "more dynamic" type than ``B``.
+
+The materialization relation is both transitive and reflexive, so it defines a
+preorder on gradual types.
+
+We also define a **consistency** relation on gradual types.
+
+A static type ``A`` is consistent with another static type ``B`` if and only if
+they are the same type (``A`` is equivalent to ``B``.)
+
+A gradual type ``A`` is consistent with a gradual type ``B``, and ``B`` is
+consistent with ``A``, if and only if ``B`` is a materialization of ``A`` or
+``A`` is a materialization of ``B``.
+
+The dynamic type ``Any`` is consistent with every type, and every type is
+consistent with ``Any``. (This must follow from the above definitions of
+materialization and consistency, but is worth stating explicitly.)
+
+The consistency relation is not transitive. ``tuple[int, int]`` is consistent
+with ``tuple[Any, int]`` and ``tuple[Any, int]`` is consistent with
+``tuple[str, int]``, but ``tuple[int, int]`` is not consistent with
+``tuple[str, int]``.
+
+The consistency relation is symmetric. If ``A`` is consistent with ``B``, ``B``
+is also consistent with ``A``. It is also reflexive: ``A`` is always consistent
+with ``A``.
+
+The consistent subtype relation
+-------------------------------
+
+Given the materialization relation and the subtyping relation, we define the
+**consistent subtype** relation over all types. A type ``A`` is a consistent
```

Comment:
A thought: "Consistent subtyping" is a central concept in the type system but the name is awkward. There are more types that are a consistent subtype of a type A than there are types that are subtypes of A, which is contrary to what I'd expect. (Someone not familiar with the theory might think that some subtypes are "consistent subtypes" and some are "inconsistent subtypes".)

What do you think of naming this relation "assignability" or even "compatibility" instead? That would be more intuitive in the rest of the spec.

Reply:
I like this. I do think it will result in more natural (and clearer to most readers not familiar with gradual typing literature) usage throughout the rest of the spec. I would like to mention the term "consistent subtyping" here (for the benefit of those familiar with the literature, but it doesn't need to be the preferred term in the spec.

I'll try a reframe of the last couple sections along these lines.

Reply:
I think I prefer "assignable to" over "compatible with." The latter (to me) suggests a symmetric relation, which consistent subtyping is not: A can be assignable to B when B is not assignable to A. This would make me tend to think "compatible with" is more like "consistent with" than "consistent subtype of." So my leaning is to avoid the term "compatible" and instead prefer "assignable to." What do you think?

Reply:
Agree "assignable" is better.

Reply:
Consistent subtyping is the awkward term from the literature.  (For historical reasons: there was first "consistency" for languages that didn't have subtyping.  Then when languages with subtyping were considered, they used "consistent subtyping" for the combination of the two.)

I do like "assignability" (as long as we're sure that it really matches what can be assigned). It's the term I use with people when I explain why we care about consistent subtyping.

By the way, I do like this more "modern" factoring of assignability into subtyping and materialization.  We can make it into a motto: assignability is subtyping plus materialization.

You can assign a value of a subtype, and you can assign a value of a more or less dynamic type.